### PR TITLE
Chat: feature to cancel active runs

### PR DESCRIPTION
**Pull request by adityaomar3**
This PR adds a cancel feature to cancel the active run on Open AI, i.e. it helps cancelling the active prompt given by user.

### CODE REVIEW

File: MAVProxy/modules/mavproxy_chat/chat_openai.py [hunk 1]
```diff
@@ -99,6 +100,23 @@ def set_api_key(self, api_key_str):
         self.assistant = None
         self.assistant_thread = None
 
+    # cancel the active run
+    def cancel_run(self):
+        # check the active thread and run
+        if (self.assistant_thread and self.latest_run is not None):
+            if (self.latest_run.status != "completed" and "cancelled"):
```
Comment (rmackay9) [inline MAVProxy/modules/mavproxy_chat/chat_openai.py:107]: Oops, I think my comment was posted somewhere else.  In any case, I think the 2nd argument looks incorrect in this "if" statement.

File: MAVProxy/modules/mavproxy_chat/chat_openai.py [hunk 2]
```diff
@@ -99,6 +100,23 @@ def set_api_key(self, api_key_str):
...
+
+                # cancel the active run
+                self.client.beta.threads.runs.cancel(
+                    thread_id=self.assistant_thread.id,
+                    run_id=self.run.id
+                )
+            else:
+                if (self.lastest_run.status == "completed"):
```
Comment (rmackay9) [inline MAVProxy/modules/mavproxy_chat/chat_openai.py:115]: little typo here, "lastest_run" should be "latest_run"

File: MAVProxy/modules/mavproxy_chat/chat_openai.py (hunk 1 above)
Comment (rmackay9) [inline MAVProxy/modules/mavproxy_chat/chat_openai.py:107]: we also need to handle the "cancelling" state

**Comment by rmackay9** [comment]
Thanks for this, I've tested and found a few little issues but once you fix those up (and re-test yourself) then I'll re-test and get this merged in

File: MAVProxy/modules/mavproxy_chat/chat_openai.py (hunk 2 above)
Comment (adityaomar3) [reply MAVProxy/modules/mavproxy_chat/chat_openai.py:115]: Oops!

File: MAVProxy/modules/mavproxy_chat/chat_openai.py (hunk 1 above)
Comment (adityaomar3) [reply MAVProxy/modules/mavproxy_chat/chat_openai.py:107]: Sure! I got it.

**Comment by adityaomar3** [comment]
HI! @rmackay9 take a look, I have updated the fixes as told, I tested it, You please take a test too, if there is any edge cases missing or changes needed.

File: MAVProxy/modules/mavproxy_chat/chat_openai.py [hunk 3]
```diff
@@ -165,7 +185,10 @@ def send_to_assistant(self, text):
             reply_messages = self.client.beta.threads.messages.list(self.assistant_thread.id,
                                                                     order="asc",
                                                                     after=input_message.id)
-            if reply_messages is None:
+
+            if (self.latest_run.status == "cancelled"):
+                return "cancelled successfully"
+            elif reply_messages is None:
```
Comment (rmackay9) [inline MAVProxy/modules/mavproxy_chat/chat_openai.py:191]: Not a blocker but the "elif" could be just an "if" because of the return one line higher

**Review by rmackay9 — APPROVED** [review]

**Comment by rmackay9** [comment]
Thanks for this.  By the way, for future PRs could you specify what testing has been done?  This is helpful for the reviewer.

I tested this and it works.  I saw one bit of odd behaviour (see below) in which I cancelled a command, "arm and takeoff to 20m" and then successfully cancelled the command but then when I put in an innocuous command a moment later ("helo" <-- was meant to be "hello") it completed the cancelled command.  I'm pretty sure this is an issue on the OpenAI side though and it didn't happen the next time.  Let's just keep an eye on this potential issue.
image

**Comment by adityaomar3** [comment]
Sure @rmackay9 thanks for the review, next time onwards I will surely add testing cases, reports from my side for a better review and tests.

Sometimes it does not cancel, maybe the reason is code latency, for some prompts(mostly in which function calling is not used) responses are generated by OpenAI very fast, while status being "in progress" when the code reaches cancel part it is already completed.

File: MAVProxy/modules/mavproxy_chat/chat_openai.py [hunk 4]
```diff
@@ -99,6 +100,25 @@ def set_api_key(self, api_key_str):
         self.assistant = None
         self.assistant_thread = None
 
+    # cancel the active run
+    def cancel_run(self):
+        # check the active thread and run
+        if (self.assistant_thread and self.latest_run is not None):
+            run_status = self.latest_run.status
+            if (run_status != "completed" and run_status != "cancelled" and
+                    run_status != "cancelling"):
```
Comment (peterbarker) [inline MAVProxy/modules/mavproxy_chat/chat_openai.py:109]:
```suggestion
            if run_status not in ["completed", "cancelled", "cancelling"]:
```

**Comment by peterbarker** [comment]
Merged, thanks!
